### ISSUE TITLE
require-atomic-updates bug

### ISSUE THREAD

**Issue by astoilkov** [eslint/eslint#12160]
**Tell us about your environment**

* **ESLint Version:** 6.2.2
* **Node Version:** 10.16.0
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**
typescript-eslint-parser but it doesn't work with the default parser as well

**Please show your full configuration:**

[ESLint Playground Link](https://eslint.org/demo#eyJ0ZXh0IjoiY29uc3QgeyByZWFkRmlsZSB9ID0gcmVxdWlyZSgnZnMnKVxuXG5hc3luYyBmdW5jdGlvbiBnZXRWYWx1ZSgpIHtcbiAgcmV0dXJuIDNcbn1cblxuYXN5bmMgZnVuY3Rpb24gZm9vKGZpbGVQYXRoLCBsYXN0TW9kaWZpZWRUaW1lKSB7XG4gIGlmIChsYXN0TW9kaWZpZWRUaW1lID09PSB1bmRlZmluZWQpIHtcbiAgICBjb25zdCB2YWx1ZSA9IGF3YWl0IGdldFZhbHVlKClcbiAgICBsYXN0TW9kaWZpZWRUaW1lID0gdmFsdWVcbiAgfVxuXG4gIHJlYWRGaWxlKGZpbGVQYXRoLCAoKSA9PiB7XG4gICAgY29uc3QgbW9kaWZpZWRUaW1lID0gbGFzdE1vZGlmaWVkVGltZSB8fCAzXG5cbiAgICByZXR1cm4gbW9kaWZpZWRUaW1lXG4gIH0pXG5cbiAgcmV0dXJuIGxhc3RNb2RpZmllZFRpbWVcbn1cblxuZm9vKClcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTAsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1lbXB0eSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtY2F0Y2giOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJuby13aXRoIjoyLCJyZXF1aXJlLWF0b21pYy11cGRhdGVzIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6Mn0sImVudiI6eyJub2RlIjp0cnVlfX19)

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const { readFile } = require('fs')

async function getValue() {
  return 3
}

async function foo(filePath, lastModifiedTime) {
  if (lastModifiedTime === undefined) {
    const value = await getValue()
    lastModifiedTime = value
  }

  readFile(filePath, () => {
    const modifiedTime = lastModifiedTime || 3

    return modifiedTime
  })

  return lastModifiedTime
}

foo()

```

**What did you expect to happen?**

Not to show an error

**What actually happened? Please include the actual, raw output from ESLint.**

Shows an error that I don't think is valid.

**Are you willing to submit a pull request to fix this bug?**

I am not familiar with ESLint codebase, so no.

**Comment by kaicataldo** [comment]
Sorry this fell through the cracks. I'm a little confused by the example code here - the `fs.readFile()` callback doesn't appear to be doing anything (and I believe would error out anyway because filePath is `undefined`). Do you have a minimal repro case that you might be able to share?

**Comment by astoilkov** [comment]
No, I don't have a repro. The example above is extracted from a real use case in our application. We have issues in two places in our codebase. I can send you both examples but they are quite complex. With the example I provided I tried to simplify the use case without including what actually the `readFile()` is doing. The call to `foo()` is quite strange indeed – I updated it to include a dummy path for more clarity.
